lib/jobsub_api.py

```diff
@@ -70,7 +86,506 @@ def jobsub_call(argv: List[str], return_output: bool = False) -> Optional[str]:
+        self.runtime = jsq_runtime_to_timedelta(runtime) if runtime else None
+        self.status = JSMAP[status] if status else None
+        self.prio = float(prio) if prio else None
+        self.size = float(size) if size else None
+        self.command = command
+
+    def _cmd(self, verbose: int, args: List[str]) -> str:
+        """code adding common arguments to commands"""
+        args.append("-G")
+        args.append(self.group)
+        if verbose:
+            args.append("--verbose")
+            args.append(str(verbose))
+        if self.pool:
+            args.append("--global-pool")
+            args.append(self.pool)
+        if self.auth_methods:
+            args.append("--auth-methods")
+            args.append(self.auth_methods)
+        if self.role:
+            args.append("--role")
+            args.append(self.role)
+        args.append(self.id)
+        rs = optfix(jobsub_call(args, True))
+        return rs
+
+    def hold(self, verbose: int = 0) -> str:
+        """Hold this job with jobsub_hold"""
+        return self._cmd(verbose, ["jobsub_hold"])
+
+    def release(self, verbose: int = 0) -> str:
+        """Release this job with jobsub_release"""
+        return self._cmd(verbose, ["jobsub_release"])
+
+    # pylint: disable=invalid-name
+    def rm(self, verbose: int = 0) -> str:
+        """Remove this job with jobsub_rm"""
+        return self._cmd(verbose, ["jobsub_rm"])
+
+    def q(self, verbose: int = 0) -> None:
+        """run 'jobsub_q' on this job and update values, status"""
+        rs = self._cmd(verbose, ["jobsub_q"])
+        lines = rs.split("\n")
+        if len(lines) == 2 and self.status is not None:
+            # we saw it previously, and now it is not showing up..
+            self.status = JobStatus.COMPLETED
+            return
+        if len(lines) > 1:
+            line = rs.split("\n")[1]
+            m = jobsub_q_re.search(line)
+            if m:
+                self.set_q_attrs(
+                    m.group("owner"),
+                    m.group("submitted"),
+                    m.group("runtime"),
+                    m.group("status"),
+                    m.group("command"),
+                    m.group("prio"),
+                    m.group("size"),
+                )
+                return
+        raise RuntimeError(f"failed jobsub_q:\n {rs}")
+
+    def q_long(self, verbose: int = 0) -> Dict[str, str]:
+        """run 'jobsub_q --long' on this job and update values, status,
+        and return the dictionary of name/value pairs from the ouput"""
+        rs = self._cmd(verbose, ["jobsub_q", "--long"])
+        lines = rs.split("\n")
+        res = {}
+        for line in lines:
+            if not line.find(" = ") > 0:
+                continue
+            k, v = line.split(" = ", 1)
+            if v[0] == '"':
+                v = v.strip('"')
+            res[k] = v
+        if len(lines) == 1 and self.status is not None:
```

Comment:
VERY nitpicky here, but I think the intent of this would be clearer if rather than checking `len(lines)`, you check to see if `rs == ""`, or even `if not rs`. This case applies to a blank result from rs, if I understand correctly, and one of those ways would communicate that clearer.

The other issue is that len(lines) COULD be 1 and the result of rs non-blank and still if the condor devs in the future forget to put line breaks in. To illustrate the point

```
>>> lines = "".split("\n")
>>> lines, len(lines)
([''], 1)
>>> lines = "blahblahoopsnolineendingkey=value".split("\n")
>>> lines, len(lines)
(['blahblahoopsnolineendingkey=value'], 1)
```

Reply:
Well, here I'm trying to handle the case that jobs just stop showing up when they complete around here; so you do
jobsub_q and it comes up empty.   When you're doing plain jobsub_q , you still get the headers, so length(lines)==2, (["JOBSUBJOBID ...", ""])l;  but for jobsub_q --long we don't get the header so length(lines)==1... So I was trying to have parallel tests in the two cases.   But you're right, at least a comment here about what the len(lines) == n means would be appropriate.

Reply:
Added a little dissertation on detecting the emtpy result, and why we're marking it "COMPLETED", in comments.

Reply:
I understand where you're coming from, but I'm still not too keen on the check counting lines, if there's a more reliable way.  How about some internal function that checks both cases?  Like a 

```python
def _jobsub_q_output_nonempty(stdout: str, long: bool) -> bool:
    # logic if long:
    if long:
        return stdout != "" 
   # logic if not long
   stuff()
```

Then you could have the same function called in both cases, so you'd satisfy having parallel checks.

Reply:
Here also switched to looking for self.id in the output.

Reply:
Sorry - I must be missing something.  Where is the `self.id` check in the output?

Reply:
Ah, I see.  You implemented the `self.id` check above in `q()`, but not here.  I like that idea to do it in both places - could you do that here as well?